### DISCUSSION TITLE
Inter-process communication

### DISCUSSION THREAD

**Discussion by KEYLEDGER** [nxrighthere/UnrealCLR#129] [General]
@nxrighthere   I'm looking for the limits of what's possible to hack with UnrealCLR. 

Q1) Strictly technically speaking, is there a way to modify what could be loaded in the host? Could a Console or a Window (WinForm or WPF) be _externally_ instantiated as long as it'd be dotnet 5? (Obviously not be present in UE rendering/Window, but spawned on the desktop/taskbar)    This would open up many interesting dev scenarios and patterns.  

Examples of scenario:
1) Controlling the app/game from a conventional UI, for demo and data visualization. Or presentation/Kiosk-like mode where UE window is on a screen and the "companion window" would be on another screen.
2) Creating custom debugging tools in the companion window. Logs, variables monitoring, flowcontrol, spawning and unloading scene/actors etc, loading dynamically .dll. etc
3) Visualization control: UE has tremendous visualization potential for science/big data/Archviz but controlling things through UE UI techs (e.g. UMG) is often not  easy or viable for such purpose. With a companion window, one could build the lab/experiment UI on it while using the UE window for visualization/simulation. 

If you think it could be done, I'd be ready to work full time on testing and developing demos for this. I believe Epic would be pleased as they are flirting more and more with those kind of non-game markets.

In case this ^ is strictly technically impossible:
A) A lib enabling an external independent app to connect to the game dll (on the same computer or remotely) to unlock as much as possible the potentials ^ described above. 
B) Such a lib ^ going as far as streaming the external app UI *inside* the game (using textures or advanced video streaming libs/algo) and sending back mouse and keyboard inputs to the external app.

What are your thoughts @nxrighthere about the technical feasibility of those scenarios?

**Comment by nxrighthere** [comment]
> Strictly technically speaking, is there a way to modify what could be loaded in the host? Could a Console or a Window (WinForm or WPF) be externally instantiated as long as it'd be dotnet 5? (Obviously not be present in UE rendering/Window, but spawned on the desktop/taskbar)

Yes, it's very easy to do. One of the options is to spawn/destroy a new process in `OnWorldBegin()`/`OnWorldEnd()` and redirect standard I/O. See [this](https://stackoverflow.com/a/697388/6836227) basic example.

You also can visualize data through web pages or use sockets, shared memory, named pipes, and so on for inter-process communication.

**Reply by KEYLEDGER** [reply]
Yes, it should be easy to redirect to a new instanced console but somehow I can't get even that to work. 
Do you have other approaches on the top of your head? Do you think there's a way to host an instance of WPF (or else) inside the game dll at runtime and communicate with it? If security context is the issue, I guess UE source could be modified to accommodate  that. Is it the issue?  

Also: Thank you for your work on UnrealCLR. Looking forward to contribute in the future.

**Reply by nxrighthere** [reply]
You can find a lot of information on the internet about [inter-process communication](https://www.google.com/search?q=C%23+IPC). From my personal experience, we often use [shared memory](https://github.com/spazzarama/SharedMemory) for high-performance data passing between two applications.
